libthymesisflow/src/thymesisflow.c

```diff
@@ -0,0 +1,243 @@
+
+        snprintf(offline_buff, sizeof(offline_buff),
+                 "/sys/devices/system/memory/memory%lu/state",
+                 probe_addr / MEMBLOCK_SIZE);
+
+        // log_info("offlining - %s\n", offline_buff);
+
+        if (access(offline_buff, W_OK) == -1) {
+
+            log_error_ext("cannot get write access- %s\n", offline_buff);
+            return -1;
+        }
+
+        FILE *fd = fopen(offline_buff, "w");
+        if (fd == NULL) {
+            log_error_ext("error opening - %s\n", offline_buff);
+            return -1;
+        }
+
+        // printf("opened - %s\n", offline_buff);
+        int res = fprintf(fd, "%s\n", offline_command);
+
+        if (res <= 0) {
+            log_error_ext("error offlining...\n");
+        }
+        // printf("written - %d\n", res);
+        fflush(fd);
+        fclose(fd);
+    }
+    return DETACH_OK;
+}
+
+int detach_memory(const char *circuit_id) {
+    connection *conn = get_conn(circuit_id);
+    if (conn == NULL) {
+        log_error_ext("error fetching connection: %s", circuit_id);
+        return ERR_MISSING_CID;
+    }
+    log_info_ext("thymesisflow - memory detach - circuit: %s\n", circuit_id);
+
+    // free memory
+    if (conn->ea == NULL) {
+        log_warn_ext(
+            "thymesisflow - cannot free memory for circuit -  %s - NULL "
+            "ref pointer \n",
+            circuit_id);
+    } else {
+        free(conn->ea);
+    }
+
+    // delete connection
+    int res_del_code;
+    // remove connection even in case of errors
+    if ((res_del_code = del_conn(circuit_id)) != CONN_SUCCESS) {
+        log_error_ext(
+            "error registering detachment for circuit %s - error: %d\n",
+            circuit_id, res_del_code);
+    }
+    return DETACH_OK;
+}
+
+int detach_compute(const char *circuit_id) {
+    connection *conn = get_conn(circuit_id);
+    if (conn == NULL) {
+        log_error_ext("error fetching circuit: %s", circuit_id);
+        return ERR_MISSING_CID;
+    }
+    log_info_ext("compute detach - circuit: %s\n", circuit_id);
+    // Check return value in detach
+
+#ifdef MOCK // assume correct detach when we mock the connection
+    int res = DETACH_OK;
+#else
+    int res = detach(conn->size);
+
+#endif
+
+    int res_del_code;
+    // remove connection even in case of errors
+    if ((res_del_code = del_conn(circuit_id)) != CONN_SUCCESS) {
+        log_error_ext(
+            "thymesisflow - error registering detachment for circuit %s "
+            "- error: %d\n",
+            circuit_id, res_del_code);
+    }
+
+    return res;
+}
+
+int attach_memory(const char *circuit_id, const char *afu_name,
+                  iport_list *ports, const uint64_t size, uint64_t *eaptr) {
+
+    connection *conn = new_conn(circuit_id, afu_name, size);
+
+    add_conn(conn);
+
+#ifdef MOCK
+    log_info_ext("mocking memory connection by allocating only 256 MB...\n");
+
+    if (posix_memalign((void **)&conn->ea, 128, (256 >> 20)) != 0) {
```

Comment:
Defining a macro for `128` could help other developers understanding what's that value about.